### ISSUE TITLE
[FEATURE_REQUEST] Children under the `RoomProvider` component should render only once initially

### ISSUE THREAD

**Issue by nimeshnayaju** [liveblocks/liveblocks#730]
I am not too familiar with the codebase, so I may be incorrect in a lot of my assumptions, so please feel free to correct me if anything seems incorrectly construed.

**Is your feature request related to a problem? Please describe.**

Yes. I noticed that all children components placed inside the `RoomProvider` component render twice initially (technically four times, but we can discount two due to React's Strict Mode behaviour). This may not be a major issue, but I wanted to start a discussion on why that may be and how we can solve it.

From what I understand after taking a look at the code, the `Room` object is created using the `useState` hook, and the code that deals with creating this `Room` object and updating it looks like the following:

https://github.com/liveblocks/liveblocks/blob/main/packages/liveblocks-react/src/factory.tsx#L172-L196

```tsx
  const [room, setRoom] = React.useState<
    Room<TPresence, TStorage, TUserMeta, TRoomEvent>
  >(() =>
    client.enter(roomId, {
      initialPresence: frozen.initialPresence,
      initialStorage: frozen.initialStorage,
      unstable_batchedUpdates: frozen.unstable_batchedUpdates,
      shouldInitiallyConnect: frozen.shouldInitiallyConnect,
    } as RoomInitializers<TPresence, TStorage>)
  );

  React.useEffect(() => {
    setRoom(
      client.enter(roomId, {
        initialPresence: frozen.initialPresence,
        initialStorage: frozen.initialStorage,
        unstable_batchedUpdates: frozen.unstable_batchedUpdates,
        withoutConnecting: frozen.shouldInitiallyConnect,
      } as RoomInitializers<TPresence, TStorage>)
    );

    return () => {
      client.leave(roomId);
    };
  }, [roomId, frozen]);
```

This may be inefficient because `RoomProvider` and all its children first render with the initially set value of `room`, and then immediately render again due to the `room` object being updated inside the `useEffect` hook. I understand the need of `entering` and `leaving` the room as the component mounts and un-mounts respectively but I think this can be done without triggering multiple renders.

Another potential issue with re-initializing `room` through the `useEffect` hook every time `roomId` changes could be that internal state of children of `RoomProvider` won't be reset when `room` changes. For example, in the sample code below, when the `roomId` value is updated, the value of `count` (if updated) in `MyComponent` doesn't reset, but the value of `shapes` does. This may be okay depending on what we believe the expected result to look like. I would, however, assume everything to be reset when the `roomId` is updated. This can still be achieved in the code block below by explicitly resetting `count` when `room` updates, but that is inefficient and we will have another double render issue.

```tsx
  function App() {
    const [roomId, setRoomId] = useState(1);
    
    return (
        <RoomProvider roomId={roomId} initialPresence={...} initialStorage={...}>
            <MyComponent />
            <button onClick={() => setRoomId(room + 1)}>Room {roomId}</button>
        </RoomProvider>
    )
}

function MyComponent() {
    const [count, setCount] = useState(0);
    const shapes = useStorage((root) => root.shapes);

    return <button onClick={() => setCount(count + 1)}>Increment</button>
}
```

**Describe the solution you'd like**

This is my very naive attempt at the solution, and please feel free to point out anything may doesn't make sense or is blatantly incorrect. And, please note that some of the functions that are called/used in the code aren't currently exposed/available.

First of all, instead of creating `room` using the `useState` hook, we could instantiate it as a mutable ref using the `useRef` hook. This will also require updating the code inside `useEffect` to connect the room and add it to client on mount, and disconnect the room and remove it from the client on unmount. 

And, in order to reset all internal states of the children of `RoomProvider`, we pass the `roomId` as the `key` property. This way, we are asking React to treat two `RoomProvider` component with different `roomIds` (as keys) as two different components that should not share any internal state. 

```tsx
function RoomProviderInner({ roomId, initialPresence, initialStorage } : { roomId: string, initialPresence: ..., initialStorage: ... }) {
    const roomRef = useRef<Room<Presence, Storage, BaseUserMeta, never>>();
  
    if (!roomRef.current) {
      roomRef.current = createInternalRoom(roomId, {
        initialPresence: initialPresence,
        initialStorage: initialStorage,
      });
    }
  
    useEffect(() => {
      const room = roomRef.current;
  
      if (!room) return;
  
      room.connect();
      client.add(room); // method not exposed/available 
  
      return () => {
        room.disconnect();
        client.remove(room); // method not exposed/available 
      };
    }, []);

    return (
      <RoomContext.Provider value={roomRef.current}>
        {children}
      </RoomContext.Provider>
    );
}

function RoomProvider({ roomId, initialPresence, initialStorage } : { roomId: string, initialPresence: ..., initialStorage: ... }) {
    return <RoomProviderInner key={roomId} roomId={roomId} initialPresence={initialPresence} .... />
}

function useRoom() {
    const room = useContext(RoomContext);
    
    if (!room) {
      throw new Error("RoomProvider not found");
    }
  
    return room;
}
```

**Comment by nvie** [comment]
Thanks for writing this issue up! Our team is going to take a look and come up with a response in the next few days! 🙏

**Comment by nvie** [comment]
Hi @nimeshnayaju, thank you for this well-written and beautifully detailed issue! 🙏 

### On the double-rendering issue

I couldn't agree more that this double-rendering behavior on mount is unintended. I've attempted to fix this in #735 in a very simple way: basically by not re-running the `setRoom` callback on the first render (because the state initializer already ran in that round). I think this should prevent the double-rendering issue. I've just published this version to NPM, under the `test` tag (`v1.0.2-test5`).

To upgrade, please run this:

```bash
$ npm install @liveblocks/core@test @liveblocks/react@test
```

Can you confirm that this fixes the bug in the way you expected? If so, I can roll this as an official 1.0.2 patch release.

### On the child components not re-rendering issue

Regarding your `<MyComponent />` child component not re-rendering: we'd need to decide if that is something we want to support or not. This example follows pretty well-defined React semantics (see for example, [this article](https://www.developerway.com/posts/react-elements-children-parents#:~:text=the%20%E2%80%9Cchildren%E2%80%9D%20pattern%20and%20a%20few%20mysteries)). So magically re-rendering all child components would not be "the React way", and as such might be unexpected or surprising to other people.

That said, I do hear you on this topic, because in applications where changing the room ID is a feature, you definitely often _want_ to reset the entire child tree. I want to take this to the team and see what we can do about this. Thanks again, and I'll let you know when I have some answers for you on the second topic!

**Comment by nimeshnayaju** [comment]
Hi @nvie, just checked out the [PR](https://github.com/liveblocks/liveblocks/pull/735), it looks very well done. Thank you for working on a fix already. From what I understand, with the fix, double rendering should not happen on production builds (where `StrictMode` is off) but still happens on local builds (where `StrictMode` is on). I tried to reproduce this through this [sandbox](https://codesandbox.io/s/hidden-rgb-3nbxp7) if you'd like to take a look. And, I understand this might not be too problematic, so please feel free to set this to low priority.

And, I definitely agree with your concerns regarding resetting the child tree when `roomId` changes. Thank you for replying with your opinions.

**Comment by nvie** [comment]
Hi @nimeshnayaju, thanks for creating that playground. I've adjusted the playground a bit to match the Liveblocks implementation a bit more closely, and I think you mean that when _changing_ the props, it re-renders twice, right? (I'm ignoring StrictMode for the moment for the sake of simplicity.)

So:
- On the first render, it used to trigger two re-renders (but PR #735 fixes that)
- But upon prop-changing the room ID, it still triggers two re-renders

I think in the current setup, that is unavoidable unfortunately. When there is an active room, it's important that the room is first left before a new room is joined, which is why it's structured like this right now. You can see the event sequence in the console logs in this screenshot:

https://codesandbox.io/s/serene-goldwasser-6l14dq?file=/src/App.tsx

Renders:
1. Joins the room + renders room 1. Click the button.
2. Prop change happens. Starts the process of leaving old and joining new room. Render room 1 (will rerender when room 2 is joined).
3. Room 2 has now been joined, rerender to reflect that.

Hope that makes sense!

**Comment by nimeshnayaju** [comment]
Hi @nvie, thanks for the correction. I do notice now that I missed the `Map` implementation when creating a room, which optimizes the extra re-render that I was seeing. Thank you for being patient about my questions. I really appreciate it. The fix works perfectly. Feel free to close the issue when you see fit.
